sql: implement row_security session variable

spilchen opened the pull request:
Complete the implementation of the `row_security` session variable, which was previously defined but non-functional.

When set to `'off'`, non-admin queries that would be affected by row-level security (RLS) fail with an error instead of silently filtering rows. This matches postgres' behavior.

Admin users and table owners (without FORCE) remain exempt from RLS regardless of this setting.

The variable defaults to `'on'` and is tracked in the memo to invalidate cached plans when changed.

Closes #138916.

Epic: none

Release note (sql change): The `row_security` session variable now behaves like in postgres, allowing users to detect when RLS is applied.

yuzefovich reviewed (approved):
Nice! Just a few nits from me, <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @yuzefovich reviewed 9 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155110#-:-Ob9-usu-q7o6JfA3QGi:bag9iea)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rytaft)

___
*[`pkg/sql/opt/memo/memo.go` line 498 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155110#-Ob9-9YPBQohIgd_e7ll:-Ob9-9YQ0LYOKxoNSWMn:bqcbykb) ([raw file](https://github.com/cockroachdb/cockroach/blob/906f4fc092480a9126493c17613d1de302472c2f/pkg/sql/opt/memo/memo.go#L498)):*
> ```Go
> 		m.disableSlowCascadeFastPathForRBRTables != evalCtx.SessionData().OptimizerDisableCrossRegionCascadeFastPathForRBRTables ||
> 		m.useImprovedHoistJoinProject != evalCtx.SessionData().OptimizerUseImprovedHoistJoinProject ||
> 		m.rowSecurity != evalCtx.SessionData().RowSecurity ||
> ```

nit: we should extend `TestMemoIsStale` for this addition.
___
*[`pkg/sql/logictest/testdata/logic_test/row_level_security` line 5766 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155110#-Ob9-i9r3-_yWtZg0w21:-Ob9-i9r3-_yWtZg0w22:b-5gswep) ([raw file](https://github.com/cockroachdb/cockroach/blob/906f4fc092480a9126493c17613d1de302472c2f/pkg/sql/logictest/testdata/logic_test/row_level_security#L5766)):*
> ```
> SET SESSION row_security = 'off';
> 
> statement error pq: query would be affected by row-level security policy for table "accounts"
> ```

nit: it's nice to also check the PG error code (if we match postgres).
___
*[`pkg/sql/vars.go` line 1688 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155110#-Ob8zuoR1wRS0KD3Sgu5:-Ob8zuoR1wRS0KD3Sgu6:b-y83fp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/906f4fc092480a9126493c17613d1de302472c2f/pkg/sql/vars.go#L1688)):*
> ```Go
> 
> 	`row_security`: {
> 		Get: func(evalCtx *extendedEvalContext, _ *kv.Txn) (string, error) {
> ```

nit: I don't immediately see why / what the difference is, but we also usually set `GetStringVal`.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql: implement row_security session variable

Complete the implementation of the `row_security` session variable,
which was previously defined but non-functional.

When set to `'off'`, non-admin queries that would be affected by
row-level security (RLS) fail with an error instead of silently
filtering rows. This matches postgres' behavior.

Admin users and table owners (without FORCE) remain exempt from RLS
regardless of this setting.

The variable defaults to `'on'` and is tracked in the memo to invalidate
cached plans when changed.

Closes #138916.

Epic: none

Release note (sql change): The `row_security` session variable now
behaves like in postgres, allowing users to detect when RLS is applied.

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155110)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft and @yuzefovich)

___
*[`pkg/sql/vars.go` line 1688 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155110#-Ob8zuoR1wRS0KD3Sgu5:-Ob9BTik2ld9oHP0Z5CT:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/906f4fc092480a9126493c17613d1de302472c2f/pkg/sql/vars.go#L1688)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I don't immediately see why / what the difference is, but we also usually set `GetStringVal`.
</blockquote></details>

Done
___
*[`pkg/sql/opt/memo/memo.go` line 498 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155110#-Ob9-9YPBQohIgd_e7ll:-Ob9BRho2RSI-HriDYIW:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/906f4fc092480a9126493c17613d1de302472c2f/pkg/sql/opt/memo/memo.go#L498)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we should extend `TestMemoIsStale` for this addition.
</blockquote></details>

Done
___
*[`pkg/sql/logictest/testdata/logic_test/row_level_security` line 5766 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155110#-Ob9-i9r3-_yWtZg0w21:-Ob9BVOD5vFMvifnE5Z0:bavalm) ([raw file](https://github.com/cockroachdb/cockroach/blob/906f4fc092480a9126493c17613d1de302472c2f/pkg/sql/logictest/testdata/logic_test/row_level_security#L5766)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: it's nice to also check the PG error code (if we match postgres).
</blockquote></details>

Done. Yes, we match postgres.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
> @yuzefovich reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155110)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)



<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm! it's nice that we're able to add this

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155110#-:-Ob9ULGx7eKLamjZIUJ7:bi5fjcn)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)



<!-- Sent from Reviewable.io -->
